src/main/fault.c

```diff
@@ -0,0 +1,800 @@
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1268.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1280.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1298.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B12AC.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B12BC.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B12D8.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B12F4.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B130C.s")
+
+extern const char* sFpExceptionNames[6];
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1328.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1340.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1354.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1368.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1374.s")
+#pragma GLOBAL_ASM("asm/us/nonmatchings/main/fault/D_800B1380.s")
+
+extern u16 sFaultFontColor;
+#endif
+
+Input sFaultInputs[4];
+s32 D_80108690;
+FaultMgr debugger;
+STACK(sFaultStack, 0x400);
+StackEntry sFaultStackInfo;
+uintptr_t D_80109090[16];
+
+void Fault_DrawRec(s32 x, s32 y, s32 w, s32 h) {
+    u16* fb = debugger.framebuf + debugger.width * y + x;
+    s32 j;
+    s32 i;
+
+    for (j = 0; j < h; j++) {
+        for (i = 0; i < w; i++) {
+            *fb++ = GPACK_RGBA5551(0, 0, 0, 1);
+        }
+        fb += debugger.width - w;
+    }
+}
+
+void Fault_PrintCharImpl(s32 x, s32 y, char fmt) {
+    u32* fontPtr = sFaultFont + (((fmt / 8) * 16) + ((fmt & 4) >> 2));
+    u16* fb = debugger.framebuf + debugger.width * y + x;
+    s32 i;
+    s32 j;
+
+    for (i = 0; i < 8; i++) {
+        u32 mask = 0x10000000 << (fmt % 4);
+        u32 font = *fontPtr;
+
+        for (j = 0; j < 8; j++) {
+            if (mask & font) {
+                *fb = sFaultFontColor;
+            }
+            mask >>= 4;
+            fb++;
+        }
+        fontPtr += 2;
+        fb += debugger.width - 8;
+    }
+    osWritebackDCacheAll();
+}
+
+void Fault_PrintChar(s32 x, s32 y, char fmt) {
+    Fault_PrintCharImpl(x, y, fmt);
+}
+
+typedef struct FaultDrawerPos {
+    /* 0x0 */ s32 x;
+    /* 0x4 */ s32 y;
+} FaultDrawerPos;
+
+void* Fault_PrintCallback(char* arg, const char* fmt, size_t count) {
+    FaultDrawerPos* pos = (FaultDrawerPos*)arg;
+
+    for (; count > 0; count--) {
+        Fault_PrintChar(pos->x, pos->y, *fmt++);
+        pos->x += 6;
+    }
+
+    return pos;
+}
+
+s32 Fault_Printf(s32 x, s32 y, const char* fmt, ...) {
+    va_list ap;
+    FaultDrawerPos pos;
+    s32 ret;
+
+    va_start(ap, fmt);
+
+    pos.x = x;
+    pos.y = y;
+    ret = _Printf((outfun*)*Fault_PrintCallback, (char*)&pos, fmt, ap);
+
+    va_end(ap);
+
+    return ret;
+}
+
+void Fault_Sleep(s32 ms) {
+    u64 cycles = (ms * osClockRate) / 1000ull;
+
+    csleep(cycles);
+}
+
+void Fault_PrintFReg(s32 x, s32 y, s32 index, f32* value) {
+    u32 raw = *(u32*)value;
```

Comment:
Type pruning has produced some issues in the past. Is there any possibility to use an union instead?
If it looks too ugly then don't worry about it

Reply:
Issue is floats are what get passed in since it is the f registers, so punning seems the best action.